File: csharp/295_Find_Medium_From_Data_Stream.cs
public class MedianFinder {
    PriorityQueue<int, int> leftQueue;
    PriorityQueue<int, int> rightQueue;

    public MedianFinder() {
        leftQueue = new PriorityQueue<int, int>(); // max heap
        rightQueue = new PriorityQueue<int, int>(); // min heap
    }
    
    public void AddNum(int num) {
        rightQueue.Enqueue(num, num);

        // remove minimum from rightQueue
        int val = rightQueue.Dequeue();
        leftQueue.Enqueue(val, -val);

        // re-balance
        if (leftQueue.Count - rightQueue.Count > 1) {
            val = leftQueue.Dequeue();
            rightQueue.Enqueue(val, val);
        }
    }
    
    public double FindMedian() {
        return leftQueue.Count > rightQueue.Count ? leftQueue.Peek() : (leftQueue.Peek() + rightQueue.Peek()) / 2.0;
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.AddNum(num);
 * double param_2 = obj.FindMedian();
 */

File: cpp/295_Find_Medium_From_Data_Stream.cpp
class MedianFinder {
public:
    MedianFinder() {

    }
    
    void addNum(int num) {
        if (left.empty() || num <= left.top()) {
            left.push(num);
        } else {
            right.push(num);
        }

        // re-balance 2 queue
        if (left.size() > right.size() + 1) {
            right.push(left.top());
            left.pop();
        }
        if(right.size() > left.size()){
            left.push(right.top());
            right.pop();
        }
    }
    
    double findMedian() {
        return left.size() > right.size() ? left.top() : (left.top() + right.top()) / 2.0;
    }
private:
    priority_queue<int> left; // contains lower value
    priority_queue<int, vector<int>, greater<int>> right; // contains higher value
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */